chore(@kadena/react-ui): Add CONTRIBUTING.md (#760)
kadena-community/kadena.js: packages/libs/react-ui/CONTRIBUTING.md

# Contributing to @kadena/react-ui

> This is an aggregation of the conventions that we have picked up while
> developing this component library. This library is still in its early stages
> development so please make any suggestions if you see ways it can be improved!

Reviewer: ```suggestion
developing this component library. This library is still in its early stages
of development so please make any suggestions if you see ways it can be improved!
```

> developing this component library. This library is still in its early stages
> development so please make any suggestions if you see ways it can be improved!

Reviewer: ```suggestion
developing this component library. This library is still in its early stages of 
development so please make any suggestions if you see ways it can be improved!
```

## Component Conventions

### Component Structure

We have had discussions around whether we want to use composition or
configuration as the pattern for developing components and have decided that
**in most cases we want to use composition, but it differs on a case by case
basis**. [Composition over Configuration][1]

#### Composition - Subcomponent structure

Composition makes sense if the component is more complex and needs a bit more
flexibility in regards to data, children, or layout. Components using this this
pattern will export one object of subcomponents that will make up the building
blocks for the entire component. These subcomponents will always export one
`Root` component as the containing component. This is inspired by [Radix
Primitives][2].

A simple example of a component that should use this pattern is the Grid
component.

```jsonc
<Grid.Root>
    <Grid.Item>{item1}</Grid.Item>
    <Grid.Item>{item2}</Grid.Item>
    <Grid.Item>{item3}</Grid.Item>
</Grid.Root>
```

The reason we opted for this approach when it comes to more complex components

> is becuase this allows keep our components more flexible by providing styled

Reviewer: ```suggestion
is because this allows us to keep our components more flexible by providing styled
```

building blocks that users can freely position based on their own unique logic.
It allows us to keep complexity outside of our component logic so we don't need
to make exceptions or combinations based on prop configuration.

#### Configuration - Configuring via props

This approach makes sense when developing simpler components. So in cases of any
`atom` or `molecule` sized components (ex. Box, Text, Button), often times using

> the subcomponent approach isn't necessary becuase the different iterations of

Reviewer: ```suggestion
the subcomponent approach isn't necessary because the different iterations of
```

how that component will look are limited and introducing more subcomponents
reduces DX.

One example of a component that we updated from the Composition to Configuration
approach was the `Button` component.

Previously when the `Button` was used with an icon, it needed to be composed
like this:

```jsonc
<Button.Root>Label<Button.Icon /></Button.Root>
```

The reason we went for this approach originally was because it matched the
pattern of the other components, however in usage it caused confusion becuase:

- It often didn't have an icon, but we still needed to use the `Root` export
- Users didn't realize the `Button` component exported it's own `Button.Icon` so
  they would use other Icon components
- It also allows for too much flexibility in terms of what can be placed within
  the button when it should only ever have text and optionally, an icon or
  loading state.

Now, it has been updated with the configuration approach and can be used in the
following way:

```jsonc
<Button icon="someIcon" iconAlign="right" loading={false}>Label</Button>
```

> This keeps the `Button` API cleaner, especially when it only houses a label, and

Reviewer: ```suggestion
This keeps the `Button` API cleaner, especially when it only uses a label, and
```

be sufficient for most components.

> TIP: Import component from @components/... in stories to check that you are
> adding new components to the components barrel file

#### Unit Tests

In some cases, there will be utility functions or additional complex logic that
you may want to test without chromatic. In these cases you can use unit tests
with Jest.

### File structure and naming naming conventions

- Always have a barrel file for every directory (every component should have an
  index.ts file)
- CSS files should be named after the component that consumes them (e.g.
  `Button.css.ts`)

### Exports

- Never export default from a file
- Always export component props with every component

## Styling with Vanilla Extract

Guidlines when styling with VE:

- Classnames should use camel-casing and always have `Class` suffix -
  `containerClass`
- You should use sprinkles whenever possible since these are predefined utility
  class. If a style value that you think is commonly used is not available as a
  sprinkle, feel free to add it.
- If you want to create multiple variations of a components style, use variants.

>   Variants should be have a `Variant` suffix - `colorVariant`

Reviewer: ```suggestion
  Variants should have a `Variant` suffix - `colorVariant`
```

cannot be applied for the sake of visual consistency. If a new style/iteration
is needed, we can discuss adding this with a designer.

> NOTE: We are starting off strict, but if necessary, we can reassess whether or
> not we want to start accepting additional styles

### Storybook, Chromatic, & Testing

We are required to create a story in storybook that showcases all functionality
for each component. This automatically gets synced with chromatic so that we
have also have regression testing. For visual related testing, this is going to
be sufficient for most components.

> TIP: import component from @components/... in stories to check that you are
> adding new components to the components barrel file

#### Unit Tests

In some cases, there will be utility functions or additional complex logic that
you may want to test without chromatic. In these cases you can use unit tests
with Jest.

### File structure and naming naming conventions

- Always have a barrel file for every directory (every component should have an
  index.ts file)
- CSS files should be named after the component that consumes them (e.g.
  `Button.css.ts`)

### Exports

> - Never export default from a file

Reviewer 2: Exception: storybook file

is because this allows us to keep our components more flexible by providing
styled building blocks that users can freely position based on their own unique
logic. It allows us to keep complexity outside of our component logic so we
don't need to make exceptions or combinations based on prop configuration.

#### Configuration - Configuring via props

This approach makes sense when developing simpler components. So in cases of any
`atom` or `molecule`<sup>\*</sup> sized components (ex. Box, Text, Button),
often times using the subcomponent approach isn't necessary because the
different iterations of how that component will look are limited and introducing
more subcomponents would negatively affect DX.

> <sup>\*</sup> These terms come from the Atomic Design methodology. If you're
> unfamiliar with this, you can read more about it
> [here](https://bradfrost.com/blog/post/atomic-web-design/) and
> [here](https://atomicdesign.bradfrost.com/chapter-2/).

One example of a component that we updated from the Composition to Configuration
approach was the `Button` component.

Previously when the `Button` was used with an icon, it needed to be composed
like this:

> ```jsonc

Reviewer 3: I thought md has support for `jsx`?

Author: I just copied it from the other file, but this is better! Updated 👍
